Tests/MongoSwiftSyncTests/MongoCursorTests.swift

```diff
@@ -0,0 +1,234 @@
+            _ = try coll.insertMany(docs)
+            let cursor = try coll.find(options: FindOptions(cursorType: .tailable))
+            expect(cursor.isAlive).to(beTrue())
+
+            let queue = DispatchQueue(label: "tailable close")
+            let allDocsLock = DispatchSemaphore(value: 1)
+
+            var allDocs: [Document]?
+            var allError: Error?
+            queue.async {
+                allDocsLock.wait()
+                defer { allDocsLock.signal() }
+                do {
+                    allDocs = try cursor.all() // should block after 3 docs are found
+                } catch {
+                    allError = error
+                }
+            }
+
+            // should be blocked on the cursor trying for more results.
+            expect(allDocsLock.wait(timeout: DispatchTime.now() + 1)).to(equal(.timedOut))
+            cursor.kill() // close cursor while it's blocking
+
+            // wait for allDocs to be updated
+            expect(allDocsLock.wait(timeout: DispatchTime.now() + 0.25)).to(equal(.success))
+            defer { allDocsLock.signal() }
+
+            expect(allError).to(beNil())
+            expect(allDocs).toNot(beNil())
+            expect(allDocs).to(equal(docs))
+        }
+    }
+
+    func testLazySequence() throws {
+        // Verify that the sequence behavior of normal cursors is as expected.
+        try self.withTestNamespace { _, _, coll in
+            try coll.insertMany([["_id": 1], ["_id": 2], ["_id": 3]])
+
+            var cursor = try coll.find()
+            expect(Array(cursor).count).to(equal(3))
+            expect(cursor.isAlive).to(beFalse())
+
+            cursor = try coll.find()
+            let mapped = Array(cursor.map { _ in 1 })
+            expect(mapped).to(equal([1, 1, 1]))
+            expect(cursor.isAlive).to(beFalse())
+
+            cursor = try coll.find()
+            let filteredMapped = cursor.filter {
+                $0.isSuccess
+            }.map { result -> Int? in
+                let document = try! result.get() // always succeeds due to filter stage
+                return document["_id"]?.asInt()
+            }
+            expect(Array(filteredMapped)).to(equal([1, 2, 3]))
+        }
+
+        // Verify that map/filter are lazy by using a tailable cursor.
+        let options = CreateCollectionOptions(capped: true, max: 3, size: 10000)
+        try self.withTestNamespace(ns: self.getNamespace(), collectionOptions: options) { _, _, coll in
+            try coll.insertMany([["_id": 1], ["_id": 2], ["_id": 3]])
+
+            let cursor = try coll.find(options: FindOptions(cursorType: .tailable))
+            // verify the cursor is lazy and doesn't block indefinitely.
+            let results = try executeWithTimeout(timeout: 1) { () -> [Int] in
+                var results: [Int] = []
+                for id in cursor.filter({ $0.isSuccess }).compactMap({ try! $0.get()["_id"]?.asInt() }) {
```

Comment:
In general, `isSuccess` could be false only in the case of the last element in the iterator, and it could be any error you see whenever communicating with the server as mentioned above. In terms of this specific test case, it should always be true except for connection errors, and in those cases, it allows the `try!` in the next stage to not panic. We could just use `try?`, but I wanted to include a `filter` here to test that it too is lazy.